src/storage/CompactionFilter.h

```diff
@@ -76,11 +79,56 @@ class StorageCompactionFilter final : public kvstore::KVFilter {
         return true;
     }
 
-    bool ttlValid(GraphSpaceID, const folly::StringPiece&) const {
-        // TODO(heng): Implement ttl filter after ttl schema pr merged in
+    bool ttlValid(GraphSpaceID spaceId, const folly::StringPiece& key,
+                  const folly::StringPiece& val) const {
+        if (NebulaKeyUtils::isVertex(key)) {
+            auto tagId = NebulaKeyUtils::getTagId(key);
+            auto schema = this->schemaMan_->getTagSchema(spaceId, tagId);
+            if (!schema) {
+                VLOG(3) << "Space " << spaceId << ", Tag " << tagId << " invalid";
+                return false;
```

Comment:
Why return false ? means this row was expired?

Reply:
When there is no schema, the data will not expire

Reply:
> When there is no schema, the data will not expire

Got , The purpose of TTL is to remove expired data. schema not found means the row is invalid data, maybe the schema has been removed.

Reply:
Yeah，compaction remove data when  `ttlValid `   function  return false.
In function `checkDataTtlValid` to check if ttl  is valid.

Anyway, ttl expired data and no valid schema data will remove